### PR TITLE
Feature: Interactive Form Dictionary (PDF 1.7 Section 12.7.2) & Appearance Streams (12.5.5)

### PR DESCRIPTION
**Pull request by tingerrr**
This PR adds the writers and types for writing the document catalog `/AcroForms` dictionary, as well as appearance streams. This PR also cleans up and unifies the API and documentation of #20, #21, #22 and #23.

It finalizes the support for interactive forms (save for signature fields), allowing work on typst to begin.

I left out signature fields because the implementation and review should happen in isolation. I want to make sure the implementation of those is sound, as it comes with a higher risk.

### CODE REVIEW

File: src/forms.rs [hunk 1]
```diff
@@ -1,6 +1,93 @@
...
+
+    // TODO: deprecated in PDF 2.0
+
+    /// Write the `/NeedAppearances` attribute to set whether to construct
+    /// appearance streams and appearance dictionaries for all widget
+    /// annotations in this document.
+    pub fn need_appearances(&mut self, need: bool) -> &mut Self {
+        self.dict.pair(Name(b"NeedAppearances"), need);
+        self
+    }
```
Comment (tingerrr) [inline src/forms.rs:28]: It was noted that this was deprecated in 2.0, I can't quite remember if it should've been included anyway.

**Comment by tingerrr** [comment]
I should also note that this has not been resolved and should be before this is merged: https://github.com/typst/pdf-writer/blob/28b79dda4d85b06a3900fde88df681db907fc0b9/src/forms.rs#L292-L300

**Comment by tingerrr** [comment]
@reknih I'll leave this in a draft to avoid that this is merged before all important points have been resolved.

**Comment by tingerrr** [comment]
Blocked by #15.

**Comment by tingerrr** [comment]
Currently, `Annotation::appearance_state` takes a `Name`, while `Field::radio_value` takes a `RadioState` which resolves to either to a custom `/Name` which is unique (or should be) across its child widget annotations, or `/Off`. I'm thinking that using `RadioState` for the annotation feels a bit like leaking the radio button impl there.

I found a few documentation errors that I fixed locally but haven't pushed yet.

Overall, I'm stuck on testing behavior, which just like the appearance is wildly different for each PDF viewer.

**Comment by tingerrr** [comment]
It appears that the optional merging of widget annotations and terminal fields (Section 12.5.6.19) is infact not optional but required by most viewers to make radio buttons work. This means that fields need an additional function that turns them into `Annotation`s.

**Comment by tingerrr** [comment]
I have rebased to incorporate #26. The following decisions must be made before merge:
- Should `/NeedAppearances` be included (see my review comment)
- What about this, I couldn't find what they mean by a character here but it's probably not the rust equivalent [pdf-writer/src/forms.rs#L292-L300](https://github.com/typst/pdf-writer/blob/28b79dda4d85b06a3900fde88df681db907fc0b9/src/forms.rs#L292-L300)
- Are we so far ok with the API, i.e. using prefixes for specializations of methods, etc.

With the changes in this PR (and #26) pdf-writer is able create PDF form fields which are accepted by the most common readers:

Full support:
- PDF Acrobat

Almost Full support:
- chromium
- firefox

Partial support:
- evince
- okular

The lack of support is seems to be a reader problem as, for some form field configurations neither forms created in accordance to the spec, not those created by Acrobat Pro behaved correctly, but had somewhat correct appearance.

**Comment by reknih** [comment]
- Did you need `NeedAppearances` for any of your form test implementations?
- I think they mean the number of bytes written, because section 7.2.2 states that "[regular characters] include bytes that are outside the ASCII character set", implying bytes are characters. This is also the reasonable interpretation that will fit the least characters. Have you investigated how readers interpret this, i.e. with four-byte (in UTF-8) CJK or emojis?
- I'm happy with the API!

An example would be good, I think.

**Comment by tingerrr** [comment]
> - Did you need NeedAppearances for any of your form test implementations?

The PDF generated by Adobe Acrobat Pro did not use it anywhere as it came with default appearances. And I would generally also provide default appearances in the typst impl later on anyway.

> - Have you investigated how readers interpret this, i.e. with four-byte (in UTF-8) CJK or emojis?

It seems to depend on what the implementation sees as a character, for chromium and firefox with `/MaxLen 1` it seems to work like this:
- 2 byte latin sequence `á` (`c3 a1`) **allowed**
- 3 byte cjk sequence `原` (`e5 8e 9f`) **allowed**
- 4 byte single emoji sequence `🏳` `(f0 9f 8f b3`) **not allowed**
- 12 byte composite emoji sequence `🏳️‍🌈` (`f0 9f 8f b3`, `ef b8 8f e2`, `80 8d f0 9f`, `8c 88`) **not allowed**

So it's neither a byte nor a codepoint nor a grapheme?

> - I'm happy with the API!

I'll add some additional methods I noticed would be handy when preparing the example.

**Comment by laurmaedje** [comment]
Is it perhaps UTF-16 len or something like that. I believe PdfDocEncoding is UTF-16.

**Comment by tingerrr** [comment]
> Is it perhaps UTF-16 len or something like that? I believe PdfDocEncoding is UTF-16.

Wouldn't that disallow a 3-byte sequence?

**Comment by reknih** [comment]
> > * Did you need NeedAppearances for any of your form test implementations?
> 
> The PDF generated by Adobe Acrobat Pro did not use it anywhere as it came with default appearances. And I would generally also provide default appearances in the typst impl later on anyway.


Okay, so we don't plan to use it, neither does Acrobat and its getting deprecated. I say drop it.

**Comment by tingerrr** [comment]
After some painful debugging, I got radio buttons to also work, I've added `Annotation::page` while testing, it's required for some annotation actions, none that we need, but we may as well add it.

Something that doesn't work however, but is an upstream issue, is that for checkboxes pdf.js will respect the `RADIOS_IN_UNISON` flag, but not for radio buttons. pdfium handles this correctly, and so does Acrobat.

Dropdowns also work as expected. The following work remains:
- [ ] Test printing
- [ ] Test more complex appearance streams
- [ ] Signature fields (likely a different PR as stated before)

You can view the new example to generate a PDF with some form fields. I'll change to use the `writer.finish()` version once I've tested everything to completion, this was easier for me to write. Feel free to test this in various readers, the feature set of forms is so large that I could've missed edge cases.

File: examples/forms.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+        // we state which appearance state each button has by default
+        annot.appearance_state(Name(b"Off"));
+
+        // most readers need appearances to distinguish the actual radio buttons
+        // the normal appearance should contain an appearance stream for each
+        // state of the button, each button should have a unique name for it's
+        // on state and it's `/Off` state, if the on-state names are not unique
+        // they will shate their state, we see this here with ch1 being in there
+        // twice
```
Comment (laurmaedje) [inline examples/forms.rs:95]: These comments would be much easier to read with some full stops in them. I would personally write all comments as complete, capitalized sentences ended with a full stop. This way, a short comment can be extended to a multi-sentence comment without it getting inconsistent.

**Comment by tingerrr** [comment]
I have a hard time to find a consistent way to get appearances to show up in most viewers, it seems they almost all support XFA forms very well, but once normal forms are used they don't properly render the appearances if at all.

Okular refuses to not show fields with their names and a simple button, evince simply shows nothing making them impossible to find on a blank page, chrome renders at least some of them correctly, firefox kind of also provides its own defaults like okular does, but they fit better with the rest.

Sadly, it seems I haven't made much progress on this, and I can't test printing if I don't see anything in the PDF first.

**Comment by epsilonhalbe** [comment]
Hey I found some time to test this on Mac:
# Review
## Mac Preview/safari/Skim
image
- radio buttons are buggy: the first radio also toggles the third one, which is usually caused by the first and third radio having the same label (I noticed that working with LaTeX)
- buttons don't have any background set - so finding them is really tricky

## firefox
image
- radio buttons work independently, but look like odd googly eyes
- the reset button appears slightly yellow on hover but is still hard to discover

## chrome/microsoft edge (on mac)
image
- radio buttons appear as rectangles and have checkmarks instead of radio button look, also first and third button go on and off together
- reset button has a border now

# Thoughts
I think that most of the difficulty with the appearance is the fallback behaviour of each pdf reader - so I think the best way would be to find a good set of default values to set like border, background color etc. to avoid having to deal with fallbacks and defaults.

Instead of digging into the PDF standard I started looking at LaTeX - bc. I know the docs are usually okay and I found [this PDF](https://mirror.ox.ac.uk/sites/ctan.org/macros/latex/contrib/hyperref/doc/hyperref-doc.pdf) documenting the hyperref package. in combination with tex.stackoverflow I was able to implement some forms in the past.

If it helps I can create some sample pdfs that have a certain behaviour/look to have a target implementation to aim for.

**Comment by tingerrr** [comment]
> radio buttons are buggy: the first radio also toggles the third one, which is usually caused by the first and third radio having the same label (I noticed that working with LaTeX)

This is intentional, I've used the RADIOS_IN_UNISON flag, meaning that those with the same appearance stream names (1 and 3) refer to the same value, the behavior of Firefox is actually buggy here.

> I think that most of the difficulty with the appearance is the fallback behaviour of each pdf reader - so I think the best way would be to find a good set of default values to set like border, background color etc. to avoid having to deal with fallbacks and defaults.

Yeah, that's what I've tried around with, but I couldn't get it to work, I assume I need to read up more on general annotation appearance.

> If it helps I can create some sample pdfs that have a certain behaviour/look to have a target implementation to aim for.

I wouldn't mind, every PDF helps to get a clearer picture!

Thank you for testing it!

**Comment by tingerrr** [comment]
I'll likely come back to this when I have more time to work on forms. Currently, typst-test and university are taking up most of my time.

**Comment by tingerrr** [comment]
The last commit adds some examples for specifying appearances, unfortunately there are far too many ways to do this to meaningfully cover them in the forms example.

While the checkboxes may not look very different, this is actually now an explicit implementation replicating the chrome default behavior (although with slightly different spacing).

Unfortunately, evince still refuses to render my checkboxes correctly, whereas the example file I looked at works well on both chrome and evince. This is basically my last pain point, I'm not sure what's missing for evince to render this correctly.

**Comment by tingerrr** [comment]
I think the example needs some polishing, what vexes me is how much redundant information has to be inserted for most viewers to accept this.

It seems, though, that the implementation itself is sufficiently complete to be able to start work on the typst side for the common field types.

I'll mark this ready for review.

File: examples/forms.rs [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    let mut content = Content::new();
+    content.save_state();
+    content.begin_text();
+    content.set_fill_gray(0.0);
+    content.set_font(font_name2, 14.0);
+    // The character 4 is a tick in this font.
+    content.show(Str(b"4"));
+    content.end_text();
+    content.restore_state();
+    pdf.form_xobject(radio_on_appearance_id, &content.finish()).bbox(bbox);
```
Comment (reknih) [inline examples/forms.rs:130]: The On Appearance is not used by Quartz / Preview on Apple platforms

https://github.com/typst/pdf-writer/assets/3874949/b79ed417-d821-4ae1-8ada-a15726ff0571

**Edit:** The issue is that the symbol font was not referenced in the content stream's own resources. I'll push this change. Now, Quartz renders an off-center tick. Pdfium / Chrome will render a tick aligned with the center bottom of the box, Acrobat will only render the tick if the mouse button is held down and pdf.js still renders a bullet. The document now passes the [3-Heights PDF Validator](https://www.pdf-online.com/osa/validate.aspx) whereas before it failed.


https://github.com/typst/pdf-writer/assets/3874949/7fdb9f1b-f8fb-4044-b3c9-2784582a0756

From Quartz with the change.

https://github.com/typst/pdf-writer/assets/3874949/2e671efd-911b-4fa2-929f-2183256c7487

From Acrobat with the change.

Comment (tingerrr) [reply examples/forms.rs]: Interesting, thanks for testing that I'm unable to test any macOS software.
